### PR TITLE
newdevice : Added additional sensors for APC IRRP100 Air Conditionner series 

### LINKED ISSUE #6855: Additional sensors for APC IRRP100 Air Conditionners series
DO NOT DELETE THIS INFORMATION.

> Please read this information carefully.

GitHub issues is for  bugs, please do not post issues asking for help or how to do X, Y or Z. 
You can use our irc channel ##librenms on freenode to ask questions or our [community site](https://community.librenms.org).
If you have a feature request please post this on our [community site](https://community.librenms.org/c/feature-requests).

Please confirm each of the sections below by putting an `x` in the box like `[x]`.

- [X] Is your install up to date? [Updating your install](http://docs.librenms.org/General/Updating/)
      Please do not submit an issue if your install is not up to date within the last 24 hours or on a stable monthly release.
- [X] Please include all of the information between the `====================================` section of `./validate.php` which you can run from the cli.
==========================================================
Component | Version
--------- | -------
LibreNMS  | 5fa8d3f64c04aef42ab3cb74089b532fbcc82285
DB Schema | 195
PHP       | 7.0.15
MySQL     | 5.5.52-MariaDB
RRDTool   | 1.4.8
SNMP      | NET-SNMP 5.7.2
==========================================================
- [X] Please provide ALL info asked for [here](http://docs.librenms.org/Support/FAQ/#faq20).
- [X] Please provide as much detail as possible.
- [X] Please do NOT post more than 10 lines of debug information here, use a pastebin service or GitHub Gists.

Discovery : https://gist.github.com/mail4git38/b8b2881c9958e4224da0fd18942d44f4
Polling : https://gist.github.com/mail4git38/47700f9927cae44eb6aa3037d5f69eea
SNMP Walk : https://gist.github.com/mail4git38/f1aa610d135c350cc205bd157211998b

I'm working on this issue. I should provide the following informations and sensors : 
- Model Number
- Serial Number
- Firmware Revision
- Operating Mode state
- Temperatures (with Thresholds)
- Humidity
- AirFlow
- Fan Speed
- Pressures
- Voltages
- Alarms sensors
- Runtimes

### PR DESCRIPTION
**Pull request by mail4git38**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [X] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

Fix #6855

### CODE REVIEW

File: includes/definitions/discovery/apc.yaml [hunk 1]
```diff
@@ -2,34 +2,294 @@ mib: PowerNet-MIB
...
+                    - { descr: onBatteryTest, graph: 0, value: 15, generic: 1 }
+                    - { descr: emergencyStaticBypass, graph: 0, value: 16, generic: 2 }
+                    - { descr: staticBypassStandby, graph: 0, value: 17, generic: 1 }
+                    - { descr: powerSavingMode, graph: 0, value: 18, generic: 0 }
+                    - { descr: spotMode, graph: 0, value: 19, generic: 0 }
+                    - { descr: eConversion, graph: 0, value: 20, generic: 0 }
+                    - { descr: chargerSpotmode, graph: 0, value: 21, generic: 0 }
+                    - { descr: inverterSpotmode, graph: 0, value: 22, generic: 0 }
+            -
+                oid: airIRRP100UnitStatusOperateMode
```
Comment (laf) [inline includes/definitions/discovery/apc.yaml]: This will go for all of the new ones you've added but you should set:

`index: 'airIRRP100UnitStatusOperateMode.{{ $index }}'`

here.

File: includes/discovery/sensors/runtime/apc.inc.php [hunk 2]
```diff
@@ -15,3 +15,65 @@
     $high_limit         = 3000;
     discover_sensor($valid['sensor'], 'runtime', $device, $oid, $index, $type, $descr, $divisor, '1', $low_limit, $low_limit_warn, $warn_limit, $high_limit, $current);
 }
+
+// InRow IRRP100
+$oids = snmp_get($device, 'airIRRP100GroupSetpointsCoolMetric.0', '-OsqnU', 'PowerNet-MIB');
```
Comment (laf) [inline includes/discovery/sensors/runtime/apc.inc.php]: Can these not be done in yaml?

File: includes/discovery/sensors/state/hirschmann.inc.php [hunk 3]
```diff
@@ -34,10 +34,10 @@
 
         foreach ($oid as $index => $entry) {
             //Discover Sensors
-            discover_sensor($valid['sensor'], 'state', $device, $cur_oid.$index, 'hmPowerSupplyStatus'.$index, $state_name, 'Power Supply '.$oid[$index]['hmPSID'], '1', '1', null, null, null, null, $oid[$index]['hmPowerSupplyStatus'], 'snmp', 'hmPowerSupplyStatus'.$index);
+            discover_sensor($valid['sensor'], 'state', $device, $cur_oid.$index, $index, $state_name, 'Power Supply '.$oid[$index]['hmPSID'], '1', '1', null, null, null, null, $oid[$index]['hmPowerSupplyStatus'], 'snmp', $index);
```
Comment (laf) [inline includes/discovery/sensors/state/hirschmann.inc.php]: Any reason why you've changed all the index values? this will cause a loss of data and shouldn't be done anyway as it's unique per sensor class.

File: includes/polling/os/apc.inc.php [hunk 4]
```diff
@@ -1,94 +1,20 @@
 <?php
 
-// PDU
-$serial = trim(snmp_get($device, 'PowerNet-MIB::rPDUIdentSerialNumber.0', '-OQv'), '"');
+$apc_serial = snmp_get_multi_oid($device, 'rPDUIdentSerialNumber.0 atsIdentSerialNumber.0 upsAdvIdentSerialNumber.0 sPDUIdentSerialNumber.0 airIRRCUnitIdentSerialNumber.0 isxModularPduIdentSerialNumber.0 airIRRP100UnitIdentSerialNumber.0 airIRRP500UnitIdentSerialNumber.0', '-OUQs', 'PowerNet-MIB');
```
Comment (laf) [inline includes/polling/os/apc.inc.php]: Have you taken this from my PR? If so, does it work ok for you?

File: includes/definitions/discovery/apc.yaml (hunk 1 above)
Comment (mail4git38) [inline includes/definitions/discovery/apc.yaml]: The index line is now added on the missing definitions.

File: includes/discovery/sensors/runtime/apc.inc.php (hunk 2 above)
Comment (mail4git38) [inline includes/discovery/sensors/runtime/apc.inc.php]: I decided to do this discovery in a dedicated file because we nned to know the alarm status (enable or not) on the runtime thresolds.

File: includes/discovery/sensors/state/hirschmann.inc.php (hunk 3 above)
Comment (mail4git38) [inline includes/discovery/sensors/state/hirschmann.inc.php]: This hirschmann file should not be part of this PR. How can I discard the commit of this file ?

File: includes/polling/os/apc.inc.php (hunk 4 above)
Comment (mail4git38) [inline includes/polling/os/apc.inc.php]: Yes I've taken this from your PR and it seems to work fine with my Air Conditionner check. I will test it with APC PDU, Automatic Transfer Switch and Netbotz devices.

File: includes/discovery/sensors/state/hirschmann.inc.php (hunk 3 above)
Comment (laf) [inline includes/discovery/sensors/state/hirschmann.inc.php]: If you edit the files and put it back to how it was then push that commit it will remove the changes.

File: includes/discovery/sensors/runtime/apc.inc.php (hunk 2 above)
Comment (laf) [inline includes/discovery/sensors/runtime/apc.inc.php]: Sorry I'm not sure what that means, the yaml will still record data like this does it's just abstracted away from writing code.

File: includes/polling/os/apc.inc.php (hunk 4 above)
Comment (laf) [inline includes/polling/os/apc.inc.php]: Ace.

File: includes/discovery/sensors/runtime/apc.inc.php (hunk 2 above)
Comment (mail4git38) [inline includes/discovery/sensors/runtime/apc.inc.php]: On the APC Air Conditionner, you can set an alert (number of weeks of runtime) on a few parts runtime (air filter, comrpessor,...). This alert can be enabled or not.
With the dedicated apc.inc.php runtime file, I just want to check if these alerts are enabled or not. If they are, we need to set the $high_limit value with the value (number of weeks converted in minutes) set in the Air Conditionner. If the alarm is disabled, the $high_limit is set to null.
I don't know if we can do this in the yaml file...

File: includes/polling/os/apc.inc.php (hunk 4 above)
Comment (mail4git38) [inline includes/polling/os/apc.inc.php]: I've tested you PR on an production environnement and here are the missing OID for an APC Netbotz 200 :

- $apc_serial : you can add emsIdentSerialNumber.0
- $apc_model : you can add emsIdentProductNumber.0
- $apc_hardware : you can add emsIdentHardwareRev.0
- $apc_version : you can add emsIdentFirmwareRev.0

Regards.

File: includes/discovery/sensors/runtime/apc.inc.php (hunk 2 above)
Comment (laf) [inline includes/discovery/sensors/runtime/apc.inc.php]: Makes sense and no, you can't do that in yaml right now.

File: includes/polling/os/apc.inc.php (hunk 4 above)
Comment (laf) [inline includes/polling/os/apc.inc.php]: Ok, add those in then :)

File: includes/discovery/sensors/runtime/apc.inc.php [hunk 5]
```diff
@@ -15,3 +15,65 @@
...
+if ($oids) {
+    echo 'APC InRow IRRP100 ';
+    // airIRRP100UnitRunHoursAirFilter
+    $index = 0;
+    $cur_oid = '.1.3.6.1.4.1.318.1.1.13.3.3.1.2.3.1.';
+    $current = snmp_get($device, 'PowerNet-MIB::airIRRP100UnitRunHoursAirFilter.'.$index, '-Oqv');
```
Comment (laf) [inline includes/discovery/sensors/runtime/apc.inc.php]: Please put the `PowerNet-MIB` at the end of the snmp_get call like:

`$current = snmp_get($device, 'airIRRP100UnitRunHoursAirFilter.'.$index, '-Oqv', 'PowerNet-MIB');`

Needs to be done for all the other calls as well.

File: includes/discovery/sensors/runtime/apc.inc.php [hunk 6]
```diff
@@ -15,3 +15,65 @@
...
+    $service_interval = snmp_get($device, 'PowerNet-MIB::airIRRP100UnitServiceIntervalAirFilter.'.$index, '-Oqv');
+    $alarm_status= snmp_get($device, 'PowerNet-MIB::airIRRP100UnitServiceIntervalAirFilterAlarm.'.$index, '-Oqv');
+    $multiplier = 60;
+    $current = ($current * 60);
```
Comment (laf) [inline includes/discovery/sensors/runtime/apc.inc.php]: You can just do `$current * $multiplier` in all of these so that changing `$multiplier` will be all that's needed (I know it won't change).

**Comment by laf** [comment]
Think this has more in than mine.

**Comment by murrant** [comment]
Oh, sorry :-)
